### ISSUE TITLE
tokenization mismatch when fine-tuning bunny-phi3

### ISSUE THREAD

**Issue by simplelifetime** [BAAI-DCAI/Bunny#111]
I'm wondering what causes this error? 
Do I have to set --version phi3 during pre-training stage? I use --version plain in pre-train stage and --version phi3 in fine-tune stage. Is this the correct setting? If so, what's causing this error?

**Comment by Isaachhh** [comment]
"I use --version plain in pre-train stage and --version phi3 in fine-tune stage." Correct.

What is your training script?

**Comment by Isaachhh** [comment]
What does `data/zkl/hf_models/phi-3` come from? Phi-3-mini-4k-instruct?

**Comment by simplelifetime** [comment]
Yes, the tokenizer_config.json looks like this:
`{
  "add_bos_token": false,
  "add_eos_token": false,
  "added_tokens_decoder": {
    "0": {
      "content": "<unk>",
      "lstrip": false,
      "normalized": false,
      "rstrip": false,
      "single_word": false,
      "special": true
    },
    "1": {
      "content": "<s>",
      "lstrip": false,
      "normalized": false,
      "rstrip": false,
      "single_word": false,
      "special": true
    },
    "2": {
      "content": "</s>",
      "lstrip": false,
      "normalized": false,
      "rstrip": true,
      "single_word": false,
      "special": false
    },
    "32000": {
      "content": "<|endoftext|>",
      "lstrip": false,
      "normalized": false,
      "rstrip": false,
      "single_word": false,
      "special": true
    },
    "32001": {
      "content": "<|assistant|>",
      "lstrip": false,
      "normalized": false,
      "rstrip": true,
      "single_word": false,
      "special": true
    },
    "32002": {
      "content": "<|placeholder1|>",
      "lstrip": false,
      "normalized": false,
      "rstrip": true,
      "single_word": false,
      "special": true
    },
    "32003": {
      "content": "<|placeholder2|>",
      "lstrip": false,
      "normalized": false,
      "rstrip": true,
      "single_word": false,
      "special": true
    },
    "32004": {
      "content": "<|placeholder3|>",
      "lstrip": false,
      "normalized": false,
      "rstrip": true,
      "single_word": false,
      "special": true
    },
    "32005": {
      "content": "<|placeholder4|>",
      "lstrip": false,
      "normalized": false,
      "rstrip": true,
      "single_word": false,
      "special": true
    },
    "32006": {
      "content": "<|system|>",
      "lstrip": false,
      "normalized": false,
      "rstrip": true,
      "single_word": false,
      "special": true
    },
    "32007": {
      "content": "<|end|>",
      "lstrip": false,
      "normalized": false,
      "rstrip": true,
      "single_word": false,
      "special": true
    },
    "32008": {
      "content": "<|placeholder5|>",
      "lstrip": false,
      "normalized": false,
      "rstrip": true,
      "single_word": false,
      "special": true
    },
    "32009": {
      "content": "<|placeholder6|>",
      "lstrip": false,
      "normalized": false,
      "rstrip": true,
      "single_word": false,
      "special": true
    },
    "32010": {
      "content": "<|user|>",
      "lstrip": false,
      "normalized": false,
      "rstrip": true,
      "single_word": false,
      "special": true
    }
  },
  "bos_token": "<s>",
  "chat_template": "{% for message in messages %}{% if message['role'] == 'system' %}{{'<|system|>\n' + message['content'] + '<|end|>\n'}}{% elif message['role'] == 'user' %}{{'<|user|>\n' + message['content'] + '<|end|>\n'}}{% elif message['role'] == 'assistant' %}{{'<|assistant|>\n' + message['content'] + '<|end|>\n'}}{% endif %}{% endfor %}{% if add_generation_prompt %}{{ '<|assistant|>\n' }}{% else %}{{ eos_token }}{% endif %}",
  "clean_up_tokenization_spaces": false,
  "eos_token": "<|endoftext|>",
  "legacy": false,
  "model_max_length": 4096,
  "pad_token": "<|endoftext|>",
  "padding_side": "left",
  "sp_model_kwargs": {},
  "tokenizer_class": "LlamaTokenizer",
  "unk_token": "<unk>",
  "use_default_system_prompt": false
}`

**Comment by Isaachhh** [comment]
Try edit [here](https://github.com/BAAI-DCAI/Bunny/blob/main/bunny/util/data_utils.py#L259) to:

```
    if conversation_lib.default_conversation.version in {"bunny", "phi3"}:
        return preprocess_bunny(sources, tokenizer, has_image=has_image)
    elif conversation_lib.default_conversation.version in {"minicpm", "llama"}:
        return preprocess_bunny_with_bos(sources, tokenizer, has_image=has_image)
```

**Comment by simplelifetime** [comment]
Thanks! That solves my problem

**Comment by Isaachhh** [comment]
@simplelifetime

This is because Microsoft upgraded Phi-3 20 days ago and modified the tokenizer about `bos_token`. ([commit](https://huggingface.co/microsoft/Phi-3-mini-4k-instruct/commit/4f818b18e097c9ae8f93a29a57027cad54b75304)).

And our code is based on the original Phi-3.
